### ISSUE TITLE
several source files are included in src.zip, but not in the database

### ISSUE THREAD

**Issue by KeuntaeShin** [github/codeql#7582]
Version: CodeQL command-line toolchain release 2.7.3.
Target language: cpp
Target source: Linux kernel

While checking the query result finding function call sites, i notice that statements of several files were not included in the result. 

After checking the database directory:

1. The src.zip includes the several files 

2. In database-create-xxx.log, i can't see traps of several files, build messages ("[build-stdout]   CC [M]  drivers/xxx.o") are shown though.

3. In database-create-xxx.log, there are error messages: 
java.io.IOException: Brotli stream decoding failed
...
org.brotli.dec.BrotliRuntimeException: Unused space
...
java.io.IOException: Brotli stream decoding failed
...
org.brotli.dec.BrotliRuntimeException: Corrupted padding bits

I feel that some source files are not treated as targets of database. Is there any way to figure out what files are handled for creating database, and are not? And why the files are not targeted?

**Comment by MathiasVP** [comment]
Hi @KeuntaeShin,

Thanks for the issue. We'll look into this, but first, we need some additional information:

- What operating system are you on?
- What SHA of the Linux kernel are you trying to build?
- What command line are you using to create the database?

Note that there should be a database of `torvalds/linux` available for you to download via https://lgtm.com/projects/g/torvalds/linux, but we recently had a technical issue which meant that we couldn't provide that database. I expect the database to be available from that link very soon again, though!

**Comment by KeuntaeShin** [comment]
Hello @MathiasVP,

1. The codeql is running on Linux xxx 5.4.0-92-generic #103-Ubuntu SMP Fri Nov 26 16:13:00 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

2. The several files are drivers and under development, so it is hard to answer. Some of driver under development are targeted as database, some of not.

3. codeql database create samsung_galaxy_s22_linux_database --language=cpp --command="[a custom build script]"

4. Due to the circumstance mentioned in second answer, it is hard to share the logs.


Please advise me to clear this problem.


Thanks,
Keuntae Shin

**Comment by KeuntaeShin** [comment]
+ Update: Another issue

import cpp

from FunctionCall fc                                                
where fc.getTarget().hasName("va_start")
select fc, fc.getEnclosingStmt(), fc.getLocation().getFile().getRelativePath()

Nothing! I could see the call sites manually in the source files though.


Thanks,

**Comment by KeuntaeShin** [comment]
The error messages have been disappeared after crafting cpp/tools/linux64/extractor forcibly using https://github.com/github/codeql/issues/7557#issuecomment-1010559878

However the several files still are not treated as targets for database.

@MathiasVP , Please feel free to share anything you have.

**Comment by MathiasVP** [comment]
> Hello @MathiasVP,
> 
> 1. The codeql is running on Linux xxx 5.4.0-92-generic # 103-Ubuntu SMP Fri Nov 26 16:13:00 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
> 2. The several files are drivers and under development, so it is hard to answer. Some of driver under development are targeted as database, some of not.
> 3. codeql database create samsung_galaxy_s22_linux_database --language=cpp --command="[a custom build script]"
> 4. Due to the circumstance mentioned in second answer, it is hard to share the logs.
> 
> Please advise me to clear this problem.
> 
> Thanks, Keuntae Shin

Thanks for the additional information. If you feel comfortable sharing the logs with GitHub Support, you can open a support ticket [here](https://support.github.com/contact) and upload the logs. They will then get routed to the correct engineering team.

> A clue that I checked:
>
> AST VIEWER in Visual Code shows nothing corresponding to the several files, while others are shown well in there.
>
> Would it relate to conditional compilation? Such as #if, #ifdef, #ifndef, #else, #elif, #endif, and defined?
> 
> Thanks,
> Keuntae Shin

Conditional compilation shouldn't be an issue, no. If the code compiles without CodeQL it _should_ also work when generating a database.

> The error messages have been disappeared after crafting cpp/tools/linux64/extractor forcibly using #7557 (comment)
> 
> However the several files still are not treated as targets for database.
>
> @MathiasVP , Please feel free to share anything if you have a clue.

I really appreciate your investigation into this! As far as we know, the issue regarding `Brotli stream decoding` errors in your original post (and in the issue described in https://github.com/github/codeql/issues/7557) is really just an unfortunate side effect of the real issue, which is that we fail to correctly compile the project. This causes us to emit some partial brotli compressed files, which we then later fail to decompress.

So the main issue is still present: we fail to build parts of the project. Having access to the logs would help us identify why we fail to build the project.

**Comment by KeuntaeShin** [comment]
@MathiasVP, Finally i found the corner case CodeQL can't handle. Would you share a below example to proper engineers for fixing it?

**Case A - Code**
```
struct a;
typedef int (*func)(struct a);
struct a {
    func f;
};

int main() {
    
}
```
**Case A - Codeql, Failed**
![image](https://user-images.githubusercontent.com/1381322/150113566-ca1e2774-fcd5-4de9-ade6-c46ad42f4f20.png)


**Case B - Code**
```
struct a;
typedef int (*func)(struct a *);
struct a {
    func f;
};

int main() {
    
}
```
**Case B - Codeql, Success**
![image](https://user-images.githubusercontent.com/1381322/150113725-bf8c0184-35ce-474c-b19d-cdd6faf076e6.png)


**Case A&B - Compile**
![image](https://user-images.githubusercontent.com/1381322/150113067-2c10cc39-c70a-44bb-9383-ea0ad8ac66d7.png)

**Comment by MathiasVP** [comment]
Thank you for this snippet, @KeuntaeShin. I've added these tests to our internal issue, and I'll keep you updated on the progress here :)

**Comment by sashabu** [comment]
@KeuntaeShin Thanks for the minimal example. We've tracked down the issue and are expecting to have the fix in 2.8.0.

**Comment by KeuntaeShin** [comment]
@MathiasVP and @sashabu , 

With 2.8.1, the issue i reported has not been observable. Thank you for your efforts.
